leveled.go

```
@@ -0,0 +1,152 @@
+/*
+   Copyright (2020) Cobalt Speech and Language Inc.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package log
+
+import (
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/cobaltspeech/log/pkg/level"
+)
+
+// LeveledLogger implements the Logger interface and uses the go stdlib log
+// package to perform logging.  Each log message has a level prefix followed by
+// JSON representation of the data being logged.  Messages with level TRACE are
+// suppressed by default and only produced if the build tag `cobalt_log_trace`
+// is used.
+type LeveledLogger struct {
+	logger      *log.Logger
+	filterLevel level.Level
+}
+
+// we define osStdErr so that it can be changed for testing
+var osStderr io.Writer = os.Stderr
+
+// NewLeveledLogger returns a new Leveledlogger that writes Error and Info
+// messages to stderr.  These defaults can be changed by providing Options.
+func NewLeveledLogger(opts ...Option) *LeveledLogger {
+	l := LeveledLogger{}
+	l.filterLevel = level.Default
+
+	for _, opt := range opts {
+		opt(&l)
+	}
+
+	if l.logger == nil {
+		l.logger = log.New(osStderr, "", log.LstdFlags)
+	}
+
+	return &l
+}
+
+type Option func(*LeveledLogger)
+
+// WithOutput returns an Option that configures the LeveledLogger to write all
+// log messages to the given Writer.  Do not combine with WithLogger
```

If you want to protect against it being used with the `WithLogger` option, you could always add some logic to the returned function to check if a logger has already been defined before setting `l.logger` (same logic could apply in the `WithLogger` option).

---

I had it before, but then NewLeveledLogger() would need to return an error and that I didn't like.